include/jemalloc/internal/lockedint.h

```diff
@@ -80,6 +80,64 @@ locked_dec_u64(tsdn_t *tsdn, malloc_mutex_t *mtx, locked_u64_t *p,
 #endif
 }
 
+static inline bool
+locked_advance_u64(tsdn_t *tsdn, malloc_mutex_t *mtx, locked_u64_t *p,
+    const uint64_t x, const uint64_t interval) {
+	LOCKEDINT_MTX_ASSERT_INTERNAL(tsdn, mtx);
+	uint64_t before, after;
+	bool overflow;
+	/*
+	 * If this function is called with a large x, extreme overflow
+	 * (after >= interval * 2) can cause interval trigger coalescing.
+	 * This is usually a desired feature, as callers would rarely wants
+	 * multiple triggers per advance.
+	 */
+#ifdef JEMALLOC_ATOMIC_U64
+	before = atomic_load_u64(&p->val, ATOMIC_RELAXED);
+	do {
+		after = before + x;
+		assert(after >= before);
+		overflow = (after >= interval);
+		if (overflow) {
+			after %= interval;
+		}
+	} while (!atomic_compare_exchange_weak_u64(&p->val, &before, after,
+	    ATOMIC_RELAXED, ATOMIC_RELAXED));
+#else
+	before = p->val;
+	after = before + x;
+	overflow = (after >= interval);
+	if (overflow) {
+		after %= interval;
+	}
+	p->val = after;
+#endif
+	return overflow;
+}
+
+static inline void
+locked_rollback_u64(tsdn_t *tsdn, malloc_mutex_t *mtx, locked_u64_t *p,
```

Comment:
Here too; maybe `locked_saturating_dec_u64`? (And a comment that we subtract as much as possible, stopping at 0).